### ISSUE TITLE
GPU not being utilized when training with "device: gpu"

### ISSUE THREAD

**Issue by chixujohnny** [microsoft/LightGBM#4475]
## Description
I already make a GPU+CUDA version of lightgbm  
BUT, no matter I use 'device':'cuda' or 'device':'gpu', the GPU seems not work at all  
And the info seems the GPU is already work.  

I use command 【htop】: CPU in a full runing status.  

I user command 【watch -n 1 nvidia-smi】: GPU  load a very small memory and really not work at all.  

It seems that no matter I use any device, LGB will only use CPU to train, and no error report.  

Train dataset almost 20G, it's not a small dataset.

**My LGB params:**
params = {
                'boosting'          : "gbdt",
                'objective'          : "rmse",
                'metric'                    : "None",
                'extra_trees'             : False,
                'num_iterations'       : 3000,
                'max_leaves'             : 31,
                'min_child_weight'   : 0.8,
                'learning_rate'     : 0.02,
                'subsample'         : 0.65,
                'colsample_bytree'  : 0.65,
                'max_bin'           : 63,
                'random_state'      : 1111,
                'first_metric_only' : True,
                'verbosity'         : 2,
                'device_type'       : "gpu",
                'gpu_platform_id'   : 0,
                'gpu_device_id'     : 0,
                'num_threads'       : 48,
                }

**Train INFO:**
[I 2021-07-19 10:19:47,261] A new study created in memory with name: no-name-8f230d34-868f-4f41-a2e4-c230727d1cc4
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Debug] Dataset::GetMultiBinFromSparseFeatures: sparse rate 0.674664
[LightGBM] [Info] Total Bins 158742
[LightGBM] [Info] Number of data points in the train set: 199555, number of used features: 2637
[LightGBM] [Info] Using GPU Device: Tesla V100S-PCIE-32GB, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 2589 dense feature groups (493.28 MB) transferred to GPU in 0.198968 secs. 1 sparse feature groups
[LightGBM] [Info] Start training from score -0.014728
Training until validation scores don't improve for 300 rounds

## Environment info

LightGBM version or commit hash:

Command(s) you used to install LightGBM

```
   cd LightGBM/python-package
   pip install -e .
```

CUDA=11.0  
Driver Version: 450.51.06
GPU: V100S

## Additional Comments

[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] CUDA currently requires double precision calculations.
[LightGBM] [Warning] Using sparse features with CUDA is currently not supported.
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] CUDA currently requires double precision calculations.
[LightGBM] [Info] LightGBM using CUDA trainer with DP float!!
[LightGBM] [Info] Total Bins 608462
[LightGBM] [Info] Number of data points in the train set: 199555, number of used features: 2636

**Comment by shiyu1994** [comment]
@chixujohnny Thanks for using LightGBM. How much memory are being occupied on GPU with the dataset? I've tested with a dataset of 400,000 training samples and 2,000 features, the GPU consumes only 1GB memory, which is reasonable.

For feature bin values, we have 400,000 * 2,000 = 800,000,000, each takes 1 byte if we use maximum bin number 256, then this will be 0.8 GB.

But this dataset gives GPU-Util close to 0%, perhaps because it is memory bounded.

We plan to release a new GPU version soon and will make the utilization of GPU better.

**Comment by chixujohnny** [comment]
> @chixujohnny Thanks for using LightGBM. How much memory are being occupied on GPU with the dataset? I've tested with a dataset of 400,000 training samples and 2,000 features, the GPU consumes only 1GB memory, which is reasonable.
> 
> For feature bin values, we have 400,000 * 2,000 = 800,000,000, each takes 1 byte if we use maximum bin number 256, then this will be 0.8 GB.
> 
> But this dataset gives GPU-Util close to 0%, perhaps because it is memory bounded.
> 
> We plan to release a new GPU version soon and will make the utilization of GPU better.

Thanks for your reply @shiyu1994 

Compare with your experiment, the using gpu memory of my job seems reasonable.

And the GPU-Util is really close to 0%. Is different from XGB 90%+ GPU-Util.  (But training speed not seems slower with XGB. So, LGB will save more power? hhhh )

I'm looking forward to your new version.

Thank you!
